src/Sema.zig

```diff
@@ -13587,7 +13587,14 @@ fn zirArrayCat(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Ai
                 const elem_default_val = if (lhs_is_tuple) lhs_ty.structFieldDefaultValue(lhs_elem_i, mod) else Value.@"unreachable";
                 const elem_val = if (elem_default_val.toIntern() == .unreachable_value) try lhs_sub_val.elemValue(mod, lhs_elem_i) else elem_default_val;
                 const elem_val_inst = Air.internedToRef(elem_val.toIntern());
-                const coerced_elem_val_inst = try sema.coerce(block, resolved_elem_ty, elem_val_inst, .unneeded);
+                const coerced_elem_val_inst = sema.coerce(block, resolved_elem_ty, elem_val_inst, .unneeded) catch |err| switch (err) {
```

Comment:
This is not the correct fix. Instead, I suggest to provide a source location rather than passing `.unneeded`.

Reply:
Yeah, so that is what @yujiri8 (https://github.com/ziglang/zig/issues/16063#issuecomment-1595912552) and I tried originally, so when instead you pass `rhs_src` and `rhs_src` in the two places, this happens:
```
$ zig-out/bin/zig test test/cases/compile_errors/invalid_array_cat_element_type.zig
test/cases/compile_errors/invalid_array_cat_element_type.zig:6:17: error: no field named 'b' in enum 'invalid_array_cat_element_type.a.Foo'
    _ = arr ++ .{.b};
               ~^~~~
test/cases/compile_errors/invalid_array_cat_element_type.zig:2:17: note: enum declared here
    const Foo = enum {
                ^~~~
test/cases/compile_errors/invalid_array_cat_element_type.zig:13:10: error: no field named 'b' in enum 'invalid_array_cat_element_type.b.Foo'
    _ = .{.b} ++ arr;
        ~^~~~
test/cases/compile_errors/invalid_array_cat_element_type.zig:9:17: note: enum declared here
    const Foo = enum {
                ^~~~
test/cases/compile_errors/invalid_array_cat_element_type.zig:18:25: error: expected type 'i32', found '*const comptime_int'
    things = things ++ .{&1};
                       ~^~~~
test/cases/compile_errors/invalid_array_cat_element_type.zig:22:15: error: expected type 'i32', found '*const comptime_int'
    things = .{&1} ++ things;
             ~^~~~
```
At first glance it might look correct but the caret isn't entirely right:
```
    things = .{&1} ++ things;
             ~^~~~
```
should be
```
    things = .{&1} ++ things;
               ^~
```
I think I tried adding a new kind of `LazySrcLoc` for this but had some problems with that as well. Hmm, I guess you should instead try to obtain a source location to `elem_val`? Maybe I figure out something later but I think I'm going to close this for now.